Doc/library/typing.rst

```diff
@@ -1230,6 +1232,103 @@ These are not used in annotations. They are building blocks for creating generic
     ``covariant=True`` or ``contravariant=True``.  See :pep:`484` for more
     details.  By default, type variables are invariant.
 
+.. class:: TypeVarTuple
+
+    :class:`Type variable <TypeVar>` tuple.
+
+    A vanilla type variable enables parameterization with a single type. A type
+    variable tuple, in contrast, allows parameterization with an
+    *arbitrary* number of types by acting like an *arbitrary* number of type
+    variables wrapped in a tuple. For example::
+
+        Shape = TypeVarTuple('Shape')
+
+        class Array(Generic[*Shape]): ...
+
+        array_1d: Array[int] = Array()       # 1 type parameter: fine
+        array_2d: Array[int, int] = Array()  # 2 type parameters: also fine
+        array_0d: Array[()] = Array()        # 0 type parameters: also fine
```

Comment:
So uhh I guess this is one of the tricky things here. The `Array` use-case was actually the specific use-case we were targeting with PEP 646 (while ensuring it was useful for other use-cases too, of course). I'm aware though that this is a relatively niche use-case, relative to the full target audience of Python.

In particular, as it stands currently, off the top of my head I can't think of any standard-library-ish examples that benefit from PEP 646. We _were_ going to add a mechanism to support `map()` and `zip()`, but we decided to omit those mechanisms until a future PEP, given that PEP 646 was already so complicated.

The options I can think of here are:
1. Leave it as it is and live with it being a bit abstract/not relevant to the full target audience.
2. Explicitly mention the context - e.g. ending the preceding paragraph with "For example, in the context of numerical computing, this enables typing of multidimensional arrays based on their shape".
3. Go with a blander example that at least doesn't run the risk of confusing non-numerical-computing readers. We could go for something as simple as `tuple[*Ts]`, just for the sake of illustrating the mechanics of how `TypeVarTuple` works (though that could be confusing because `tuple` is already kinda variadic in the first place). Otherwise, we could just do something like `class Spam(Generic[*Ts]): ...; Spam[int]; Spam[int, str]`.
4. Brainstorm for a while longer to see if we can come up with a class-like object other than `Array` that would benefit from being variadic.

For what it's worth, though, my hunch is that most of the major use cases of variadic generics are relatively niche in the first place. At least, that's my vague memory from when I tried to go through all the issues asking for variadic generics in Python last year, pointing them to the PEP 646 review so they could check whether it would work for their use-case.

Reply:
You can actually do a pretty decent job of stubbing `functools.partial` with PEP 646, which is pretty rad:

```python
from typing import Any, Callable, Concatenate, Generic, ParamSpec, TypeVar, TypeVarTuple, overload

_P1 = ParamSpec("_P1")
_P2 = ParamSpec("_P2")
_T = TypeVar("_T")
_R_co = TypeVar("_R_co", covariant=True)
_Ts = TypeVarTuple("_Ts")

class partial(Generic[_P1, _P2, _T, _R_co, *_Ts]):
    @overload
    def __new__(cls, __func: Callable[_P1, _R_co]) -> partial[_P1, _P1, Any, _R_co]: ...
    @overload
    def __new__(cls, __func: Callable[Concatenate[*_Ts, _P2], _R_co], *args: *_Ts) -> partial[Concatenate[*_Ts, _P2], _P2, Any, _R_co, *_Ts]: ...
    @overload
    def __new__(cls, __func: Callable[_P1, _R_co], *args: *_Ts, **kwargs: _T) -> partial[_P1, ..., _T, _R_co, *_Ts]: ...
    def __call__(self, *args: _P2.args, **kwargs: _P2.kwargs) -> _R_co: ...
    func: Callable[_P1, _R_co]
    args: tuple[*_Ts]
    keywords: dict[str, _T]
```

But that's almost certainly too complex for a docs example...

Reply:
What about something like the use case discussed here -- decorators which only accept positional args? https://github.com/python/typing/issues/1000

Reply:
A non-trivial example that cannot be typed correctly without `Ts`:

```
def call_soon(callback, *args):
     ...
    callback(*args)
```

```
def call_soon(callback: Callable[[*Ts], None], *args: *Ts) -> None:
    ...
```

Reply:
Hah, glad you two are more creative on this one than I am :)

I'd hesitate to use `call_soon` as the first example, because it also involves `*args: *Ts`. But I've swapped it in for the `*args: *Ts` example - it's much better than `args_to_tuple`.

Alex: good point re `partial`, but agreed that it's probably a bit too complex for these docs. But Pradeep, `remove_first_element` looks good. One downside is that it's a bit harder to emphasise the "This allow a generic object to take an arbitrary number of type parameters" side of it, as opposed to "A TypeVarTuple can be _bound_ to an arbitrary number of type arguments", which feels slightly different - but I think it's probably a reasonable tradeoff for the increased simplicity of this first example.